ocean_lib/assets/v4/asset.py

```diff
@@ -0,0 +1,244 @@
+        metadata: Optional[dict] = None,
+        services: Optional[list] = None,
+        credentials: Optional[dict] = None,
+        nft: Optional[dict] = None,
+        datatokens: Optional[list] = None,
+        event: Optional[dict] = None,
+        stats: Optional[dict] = None,
+    ) -> None:
+
+        self.did = did
+        self.context = context or ["https://w3id.org/did/v1"]
+        self.metadata = metadata
+        self.version = "4.0.0"
+        self.services = services or []
+        self.credentials = credentials or {}
+        self.nft = nft
+        self.datatokens = datatokens
+        self.event = event
+        self.stats = stats
+
+        network_url = os.getenv("OCEAN_NETWORK_URL")
+        if chain_id:
+            self.chain_id = chain_id
+        elif network_url:
+            w3 = get_web3(network_url)
+            self.chain_id = w3.eth.chain_id
+        else:
+            raise TypeError(f"The chain ID is not provided.")
+
+    @property
+    def requires_address_credential(self) -> bool:
+        """Checks if an address credential is required on this asset."""
+        manager = AddressCredential(self)
+        return manager.requires_credential()
+
+    @property
+    def allowed_addresses(self) -> list:
+        """Lists addresses that are explicitly allowed in credentials."""
+        manager = AddressCredential(self)
+        return manager.get_addresses_of_class("allow")
+
+    @property
+    def denied_addresses(self) -> list:
+        """Lists addresses that are explicitly denied in credentials."""
+        manager = AddressCredential(self)
+        return manager.get_addresses_of_class("deny")
+
+    def add_address_to_allow_list(self, address: str) -> None:
+        """Adds an address to allowed addresses list."""
+        manager = AddressCredential(self)
+        manager.add_address_to_access_class(address, "allow")
+
+    def add_address_to_deny_list(self, address: str) -> None:
+        """Adds an address to the denied addresses list."""
+        manager = AddressCredential(self)
+        manager.add_address_to_access_class(address, "deny")
+
+    def remove_address_from_allow_list(self, address: str) -> None:
+        """Removes address from allow list (if it exists)."""
+        manager = AddressCredential(self)
+        manager.remove_address_from_access_class(address, "allow")
+
+    def remove_address_from_deny_list(self, address: str) -> None:
+        """Removes address from deny list (if it exists)."""
+        manager = AddressCredential(self)
+        manager.remove_address_from_access_class(address, "deny")
+
+    @classmethod
+    def from_dict(cls, dictionary: dict) -> "V4Asset":
+        """Import a JSON dict into this Asset."""
+        values = copy.deepcopy(dictionary)
+        id_key = "id" if "id" in values else "_id"
+        did = values.pop(id_key)
+        chain_id = values.pop("chainId")
+        context = values.pop("@context")
+
+        if "metadata" in values:
```

Comment:
I think you can remove the assignment part, just use these pops in the cls() call. Except for services, of course, which needs some cajoling before sending it to the constructor :)